esinstall/src/entrypoints.ts

```diff
@@ -0,0 +1,337 @@
+      return browserField;
+    } else if (typeof browserField === 'object') {
+      let browserEntrypoint =
+        (entry && browserField[entry]) ||
+        browserField['./index.js'] ||
+        browserField['./index'] ||
+        browserField['index.js'] ||
+        browserField['index'] ||
+        browserField['./'] ||
+        browserField['.'];
+
+      if (typeof browserEntrypoint === 'string') {
+        return browserEntrypoint;
+      }
+    }
+  }
+
+  // If browser object is set but no relevant entrypoint is found, fall back to "main".
+  return manifest.main;
+}
+
+/**
+ * Given an ExportMapEntry find the entry point, resolving recursively.
+ */
+export function findExportMapEntry(
+  exportMapEntry?: ExportMapEntry,
+  conditions?: string[],
+): string | undefined {
+  // If this is a string or undefined we can skip checking for conditions
+  if (typeof exportMapEntry === 'string' || typeof exportMapEntry === 'undefined') {
+    return exportMapEntry;
+  }
+
+  let entry = exportMapEntry;
+  if (conditions) {
+    for (let condition of conditions) {
+      if (entry[condition]) {
+        entry = entry[condition];
+      }
+    }
+  }
+
+  return (
+    findExportMapEntry(entry?.browser) ||
+    findExportMapEntry(entry?.import) ||
+    findExportMapEntry(entry?.default) ||
+    findExportMapEntry(entry?.require) ||
+    undefined
+  );
+}
+
+type ResolveEntrypointOptions = {
+  cwd: string;
+  packageLookupFields: string[];
+};
+
+/**
+ * Resolve a "webDependencies" input value to the correct absolute file location.
+ * Supports both npm package names, and file paths relative to the node_modules directory.
+ * Follows logic similar to Node's resolution logic, but using a package.json's ESM "module"
+ * field instead of the CJS "main" field.
+ */
+export function resolveEntrypoint(
+  dep: string,
+  {cwd, packageLookupFields}: ResolveEntrypointOptions,
+): string {
+  // We first need to check for an export map in the package.json. If one exists, resolve to it.
+  const [packageName, packageEntrypoint] = parsePackageImportSpecifier(dep);
+  const [packageManifestLoc, packageManifest] = resolveDependencyManifest(packageName, cwd);
+
+  if (packageManifestLoc && packageManifest && typeof packageManifest.exports !== 'undefined') {
+    const exportField = (packageManifest as PackageManifestWithExports).exports;
+
+    // If this is a non-main entry point
+    if (packageEntrypoint) {
+      const normalizedMap = explodeExportMap(exportField, {
+        cwd: path.dirname(packageManifestLoc),
+      });
+      const mapValue = normalizedMap && Reflect.get(normalizedMap, './' + packageEntrypoint);
+
+      if (typeof mapValue !== 'string') {
+        throw new Error(
+          `Package "${packageName}" exists but package.json "exports" does not include entry for "./${packageEntrypoint}".`,
+        );
+      }
+      return path.join(packageManifestLoc, '..', mapValue);
+    } else {
+      const exportMapEntry = exportField['.'];
```

Comment:
Here, too, do we need to handle the scenario where

```json
  "exports": { "browser": "…" }
```

?

Reply:
Added this.